### ISSUE TITLE
[BUG] Possible issue with downloads

### ISSUE THREAD

**Issue by cnishina** [microsoft/playwright#1967]
- Playwright Version: 0.15.0
- Operating System: Mac

TL;DR: downloads work. Just not all the time and I was trying to debug why.

**Downloading on file.io**

Upload a file to https://file.io as a good way to test this feature. Once you click on the link of an uploaded file, the file downloads and the link no longer works.

```javascript
// starting up chromium (also tried firefox) with accept downloads
const headless = false;
const playwrightLaunch = await playwright['chromium'].launch({headless});
const context = await playwrightLaunch.newContext({acceptDownloads: true});
page = await context.newPage();
await page.goto('https://file.io', {waitUntil: 'networkidle'});

// created a file for upload
let filename = path.resolve(__dirname, 'hello.zzz');
fs.writeFileSync(filename, 'foo');
const input = await page.$('input[name="file"]');
await input.setInputFiles(filename);

// tried to download. 
let [ download1 ] = await Promise.all([
  page.waitForEvent('download'),
  page.click('span.lead a[target="_blank"]')
]);
console.log(await download1.path());
```
This times out. Also added some debugging to crPage.js:

```javascript
_addSessionListeners() {
    this._eventListeners = [
        // ...
        helper_1.helper.addEventListener(this._client, 'Page.downloadProgress', event => {
            console.log('event: ' + JSON.stringify(event));
            return this._onDownloadProgress(event);
        }),
        // ...
```

The output of my test was a total byte of 0. This is never updated.

```
_onDownloadWillBegin
event: {"guid":"05ea81aa-7849-4308-9b96-6510ed05fb1d","totalBytes":0,"receivedBytes":0,"state":"inProgress"}
_onDownloadProgress inProgress
{"guid":"05ea81aa-7849-4308-9b96-6510ed05fb1d","totalBytes":0,"receivedBytes":0,"state":"inProgress"}
event: {"guid":"05ea81aa-7849-4308-9b96-6510ed05fb1d","totalBytes":0,"receivedBytes":11,"state":"inProgress"}
_onDownloadProgress inProgress
{"guid":"05ea81aa-7849-4308-9b96-6510ed05fb1d","totalBytes":0,"receivedBytes":11,"state":"inProgress"}
```

My guess is that there is some logic happening that if the totalBytes != receivedBytes that the state is inProgress.

**Downloading from GitHub**

```
page = await context.newPage();
page.goto('https://github.com/microsoft/playwright/releases', {waitUntil: 'networkidle'});

let [ download2 ] = await Promise.all([
  page.waitForEvent('download'),
  page.click('a[href="/microsoft/playwright/archive/v0.13.0.zip"]')
]);

console.log(await download2.path());
```
Output:
```bash
path /var/folders/ft/1c7_rwcx16xcckqq1sl43b2m0000gn/T/playwright_downloads-Ql10tO
event: {"guid":"80c57044-121f-4819-aa4f-9d0477be0305","totalBytes":0,"receivedBytes":0,"state":"inProgress"}
_onDownloadProgress inProgress
{"guid":"80c57044-121f-4819-aa4f-9d0477be0305","totalBytes":0,"receivedBytes":0,"state":"inProgress"}
event: {"guid":"80c57044-121f-4819-aa4f-9d0477be0305","totalBytes":0,"receivedBytes":0,"state":"inProgress"}
_onDownloadProgress inProgress
{"guid":"80c57044-121f-4819-aa4f-9d0477be0305","totalBytes":0,"receivedBytes":0,"state":"inProgress"}
event: {"guid":"80c57044-121f-4819-aa4f-9d0477be0305","totalBytes":0,"receivedBytes":1573114,"state":"inProgress"}
_onDownloadProgress inProgress
{"guid":"80c57044-121f-4819-aa4f-9d0477be0305","totalBytes":0,"receivedBytes":1573114,"state":"inProgress"}
event: {"guid":"80c57044-121f-4819-aa4f-9d0477be0305","totalBytes":0,"receivedBytes":1573114,"state":"inProgress"}
_onDownloadProgress inProgress
{"guid":"80c57044-121f-4819-aa4f-9d0477be0305","totalBytes":0,"receivedBytes":1573114,"state":"inProgress"}
event: {"guid":"80c57044-121f-4819-aa4f-9d0477be0305","totalBytes":1573114,"receivedBytes":1573114,"state":"inProgress"}
_onDownloadProgress inProgress
{"guid":"80c57044-121f-4819-aa4f-9d0477be0305","totalBytes":1573114,"receivedBytes":1573114,"state":"inProgress"}
event: {"guid":"80c57044-121f-4819-aa4f-9d0477be0305","totalBytes":1573114,"receivedBytes":1573114,"state":"completed"}
_onDownloadProgress completed
{"guid":"80c57044-121f-4819-aa4f-9d0477be0305","totalBytes":1573114,"receivedBytes":1573114,"state":"completed"}
_reportFinished
```

**Describe the bug**

This is possibly a bug or this is working as-intended. If this is as-intended, then there should be documentation what situation why. When downloading, files, I sometimes will do a HEAD call on the file, that provides me a content-length header. When I do it for GitHub, I get a content-length of 1573114 (`curl -v -X HEAD https://codeload.github.com/microsoft/playwright/zip/v0.13.0`). When I do it for file.io, it invalidates the download. Anyway, I did make a HEAD call against the link, there is no content-length. Anyway, I think it has something to do with that.

Example test code: https://github.com/cnishina/protractor-experiments/tree/master/playwright-files

**Comment by cnishina** [comment]
I feel very unlucky. 😭

**Comment by rwoll** [comment]
I had been running into a similar problem (that I wasn't quite able to work around). It turned out the anchors I was clicking had **`target="_blank"`** set so the download, while logically associated with the page I was clicking on, was actually associated with a new target/page.

There might be other things going on here, but it looks like the two links from your examples above have slightly different attributes (including **`target="_blank"`** in the case that's not working) so you may be running into the same thing:

GitHub Zip download link:

```html
<a href="/microsoft/playwright/archive/v0.13.0.zip" rel="nofollow">…</a>
```

file[.]io download link

```
<a target="_blank" href="h…">…</a>
```

If that's the case, hopefully there is a suggested work around or a potential fix. (From looking at the docs and the source, there's not a 'download' event at the Browser and Context level exposed by Playwright, but that may be able to help in this case.)

**Comment by cnishina** [comment]
Interesting. I'll try to hack in to remove the `target="_blank"`.

**Comment by pavelfeldman** [comment]
Oh, that's a great investigation, thanks @cnishina , @rwoll . I don't think we have tests for downloads that also specify other targets.

**Comment by yury-s** [comment]
Thanks for the investigation! This is similar to the other new page downloads that we currently don't report correctly (e.g. [alt+click](https://github.com/microsoft/playwright/blob/master/test/download.spec.js#L99)). I've added a new test for this specific case #1984

**Comment by pavelfeldman** [comment]
This test has been fixed by now, so closing this bug. Please let us know if that did not help via reopening or filing a new one!
